### PR TITLE
Refactor subgraph merging

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44995 Clean up sortReverseTopological copy pasta
* #44972 [JIT] Regularize tensorexpr fuser strategy with other fusers
* **#44238 Refactor subgraph merging**

Refactor create_autodiff_subgraphs to use the same updating of output aliasing properties logic as tensorexpr fuser, and factor that out to a common function in subgraph utils.

The first version of this PR tried to reuse the Value Map implementation, but ran into issues where we were using as a key the address of a Value who had been destroy'd, and so when we allocated a new Value it's key would be erased. Even though the "Value Mapper" approach is "dumb" in that we're refiguring out information we could try to maintain throughout the transformations, to me it's simpler and less error prone. 

Differential Revision: [D23871565](https://our.internmc.facebook.com/intern/diff/D23871565)

### CODE REVIEW

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 1]
```diff
@@ -9,6 +10,82 @@ bool hasSubgraph(Node* n) {
...
+    }
+    placeholder_node_->destroy();
+  }
+
+  std::vector<c10::optional<const Use>> last_uses_;
+  size_t subgraph_num_outputs_;
+  Node* placeholder_node_;
+};
+
+Node* executeSubgraphMergeAndUpdateAliasing(
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: is it really worth it outlining `executeSubgraphMergeAndUpdateAliasing` ?

File: torch/csrc/jit/passes/utils/subgraph_utils.h [hunk 2]
```diff
@@ -26,6 +27,13 @@ TORCH_API Node* createSingletonSubgraph(
     Symbol subgraphKind,
     std::unordered_map<Value*, Value*>& vmap);
 
+// Creates a new subgraph that only contains `n`, amd udpates the new outputs
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/subgraph_utils.h:30]: amd -> and

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: it has two call sites

Comment (Krovatkin) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: yeah, my point was that it's a 3-line helper that requires callers to create lambdas that incidentally also capture the whole world (I can't easily tell what arguments those lambdas need). This makes the logic harder to follow. that's just my 2 cents.

Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: How would you refactor it ?

Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: I suppose I could just make it take in the two node ptrs and infer if we’re creating a new subgraph or merging

Comment (Krovatkin) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: I'd **consider** (not suggesting to actually do it) inlining it unless we think that `executeSubgraphMergeAndUpdateAliasing` might change frequently.

Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:73]: Im gonna land to not re-incur wraath of CI but i might take a look at this later, as part of another PR/post release rush

**Pull request closed without merging**